sql/catalyst/src/main/java/org/apache/spark/sql/catalog/v2/expressions/Expressions.java

```diff
@@ -0,0 +1,141 @@
+ */
+@Experimental
+public class Expressions {
+  private Expressions() {
+  }
+
+  /**
+   * Create a logical transform for applying a named transform.
+   * <p>
+   * This transform can represent applying any named transform.
+   *
+   * @param name the transform name
+   * @param args expression arguments to the transform
+   * @return a logical transform
+   */
+  public Transform apply(String name, Expression... args) {
+    return LogicalExpressions.apply(name, args);
+  }
+
+  /**
+   * Create a named reference expression for a column.
+   *
+   * @param name a column name
+   * @return a named reference for the column
+   */
+  public NamedReference column(String name) {
+    return LogicalExpressions.reference(name);
+  }
+
+  /**
+   * Create a literal from a value.
+   *
+   * @param value a value
+   * @param <T> the JVM type of the value
+   * @return a literal expression for the value
+   */
+  public <T> Literal<T> literal(T value) {
+    return LogicalExpressions.literal(value);
+  }
+
+  /**
+   * Create a bucket transform for one or more columns.
+   * <p>
+   * This transform represents a logical mapping from a value to a bucket id in [0, numBuckets)
+   * based on a hash of the value.
+   *
+   * @param numBuckets the number of output buckets
+   * @param columns input columns for the bucket transform
+   * @return a logical bucket transform
+   */
+  public Transform bucket(int numBuckets, String... columns) {
+    return LogicalExpressions.bucket(numBuckets, columns);
+  }
+
+  /**
+   * Create an identity transform for a column.
+   * <p>
+   * This transform represents a logical mapping from a value to itself.
+   *
+   * @param column an input column
+   * @return a logical identity transform
+   */
+  public Transform identity(String column) {
+    return LogicalExpressions.identity(column);
+  }
+
+  /**
+   * Create a year transform for a timestamp or date column.
+   * <p>
+   * This transform represents a logical mapping from a timestamp or date to a year, such as 2018.
+   *
+   * @param column an input timestamp or date column
+   * @return a logical year transform
+   */
+  public Transform year(String column) {
+    return LogicalExpressions.year(column);
+  }
+
+  /**
+   * Create a month transform for a timestamp or date column.
+   * <p>
+   * This transform represents a logical mapping from a timestamp or date to a month, such as
+   * 2018-05.
+   *
+   * @param column an input timestamp or date column
+   * @return a logical month transform
+   */
+  public Transform month(String column) {
+    return LogicalExpressions.month(column);
+  }
+
+  /**
+   * Create a date transform for a timestamp or date column.
+   * <p>
+   * This transform represents a logical mapping from a timestamp or date to a date, such as
+   * 2018-05-13.
+   *
+   * @param column an input timestamp or date column
+   * @return a logical date transform
+   */
+  public Transform date(String column) {
```

Comment:
what are the timezone semantics for all of these datetime functions?

Reply:
If this is used to pass dates to Spark, then values must be the date in UTC. This corresponds to the UTC value passed to the data source when writing. The exact semantics here will be documented when we support passing values to Spark, which is a future extension.

When this is used as a hidden transform and values are not passed to Spark, it doesn't matter. The current use simply tells the data source to break the data into daily partitions (or yearly, or monthly, or hourly). As long as those are regular intervals and the source knows how to prune them, it can use whatever representation it chooses to.

Reply:
I should also note the suggestion I made above for using transforms to pass data back to Spark. I think we should pass back `builtin.date` as the transform so that Spark knows the date is in its internal format.